Seems freezing occurs randomly. I ran 3 additional `gmake testall`. 2 passed vs 1 froze.

8 core on my machine, and the latest output message in each frozen run is `cmdlineargs` always. 
```julia
channels (3)              |   14.01  |  1.10  |  7.8 | 347.42     |  0.81
misc (6)                  |   48.05  |  2.33  |  4.8 | 1514.21    |  0.67
libgit2 (7)               |   67.62  |  0.71  |  1.1 | 452.90     |  1.21
ranges (4)                |   92.71  |  4.86  |  5.2 | 7165.24    |  0.71
examples (9)              |   27.80  |  0.38  |  1.4 | 614.00     |  0.80
distributed (8)           |  113.42  |  0.02  |  0.0 | 20.62      |  0.56
cmdlineargs (2)           |  142.23  |  0.31  |  0.2 | 71.36      |  0.55
^C
signal (2): Interrupt
while loading /usr/home/iblis/git/julia/test/runtests.jl, in expression starting on line 29
__sys_kevent at /lib/libc.so.7 (unknown line)
__thr_kevent at /usr/src/lib/libthr/thread/thr_syscalls.c:382
uv__io_poll at /usr/home/iblis/git/julia/deps/srccache/libuv-52d72a52cc7ccd570929990f010ed16e2ec604c8/
src/unix/kqueue.c:160
uv_run at /usr/home/iblis/git/julia/deps/srccache/libuv-52d72a52cc7ccd570929990f010ed16e2ec604c8/src/u
nix/core.c:354
process_events at ./libuv.jl:82 [inlined]
wait at ./event.jl:216
wait at ./event.jl:27
jlcall_wait_19639 at /usr/home/iblis/git/julia/usr/lib/julia/sys.so (unknown line)
wait at ./task.jl:181
jlcall_wait_19633 at /usr/home/iblis/git/julia/usr/lib/julia/sys.so (unknown line)
sync_end at ./task.jl:274
macro expansion at ./task.jl:303 [inlined]
#40 at /usr/home/iblis/git/julia/test/runtests.jl:49
cd at ./file.jl:70
unknown function (ip: 0x80229ac73)
do_call at /usr/home/iblis/git/julia/src/interpreter.c:66
eval at /usr/home/iblis/git/julia/src/interpreter.c:246
eval_body at /usr/home/iblis/git/julia/src/interpreter.c:544
jl_interpret_toplevel_thunk at /usr/home/iblis/git/julia/src/interpreter.c:697
jl_toplevel_eval_flex at /usr/home/iblis/git/julia/src/toplevel.c:597
jl_parse_eval_all at /usr/home/iblis/git/julia/src/ast.c:908
jl_load at /usr/home/iblis/git/julia/src/toplevel.c:620 [inlined]
jl_load_ at /usr/home/iblis/git/julia/src/toplevel.c:627
include_from_node1 at ./loading.jl:556
jlcall_include_from_node1_18748 at /usr/home/iblis/git/julia/usr/lib/julia/sys.so (unknown line)
include at ./sysimg.jl:14
jlcall_include_18691 at /usr/home/iblis/git/julia/usr/lib/julia/sys.so (unknown line)
process_options at ./client.jl:315
_start at ./client.jl:383
jlcall__start_18926 at /usr/home/iblis/git/julia/usr/lib/julia/sys.so (unknown line)
jl_stderr_obj at /usr/home/iblis/git/julia/usr/bin/julia (unknown line)
jl_stderr_obj at /usr/home/iblis/git/julia/usr/bin/julia (unknown line)
jl_stderr_obj at /usr/home/iblis/git/julia/usr/bin/julia (unknown line)
unknown function (ip: 0x800624fff)
unknown function (ip: 0xffffffffffffffff)
Allocations: 2349656 (Pool: 2348919; Big: 737); GC: 5
WARNING: ErrorException
atexit hook threw an error: ErrorException("schedule: Task not runnable")gmake[1]: *** [Makefile:     18: all] Error 130
```

Second frozen run:
```julia
...
WARNING: OutOfMemoryError
atexit hook threw an error: OutOfMemoryError()gmake[1]: *** [Makefile:18: all] Error 130
gmake: *** [Makefile:562: testall] Interrupt
```

--- a/base/repl/LineEdit.jl
+++ b/base/repl/LineEdit.jl
@@ -186,7 +186,7 @@ end
 prompt_string(s::PromptState) = prompt_string(s.p)
 prompt_string(p::Prompt) = prompt_string(p.prompt)
 prompt_string(s::AbstractString) = s
-prompt_string(f::Function) = eval(Expr(:call, f))
+prompt_string(f::Function) = Base.invokelatest(f)
 
 refresh_multi_line(s::ModeState) = refresh_multi_line(terminal(s), s)
 refresh_multi_line(termbuf::TerminalBuffer, s::ModeState) = refresh_multi_line(termbuf, terminal(s), s)
@@ -626,8 +626,8 @@ default_enter_cb(_) = true
 
 write_prompt(terminal, s::PromptState) = write_prompt(terminal, s.p)
 function write_prompt(terminal, p::Prompt)
-    prefix = isa(p.prompt_prefix,Function) ? eval(Expr(:call, p.prompt_prefix)) : p.prompt_prefix
-    suffix = isa(p.prompt_suffix,Function) ? eval(Expr(:call, p.prompt_suffix)) : p.prompt_suffix
+    prefix = prompt_string(p.prompt_prefix)
+    suffix = prompt_string(p.prompt_suffix)
     write(terminal, prefix)
     write(terminal, Base.text_colors[:bold])
     write(terminal, prompt_string(p.prompt))

--- a/test/repl.jl
+++ b/test/repl.jl
@@ -535,46 +535,43 @@ end
 
 # Simple non-standard REPL tests
 if !Sys.iswindows() || Sys.windows_version() >= Sys.WINDOWS_VISTA_VER
-    for prompt = ["testπ", () -> randstring(rand(1:10))]
-        stdin_write, stdout_read, stdout_read, repl = fake_repl()
-        panel = LineEdit.Prompt(prompt;
-            prompt_prefix="\e[38;5;166m",
-            prompt_suffix=Base.text_colors[:white],
-            on_enter = s->true)
+    stdin_write, stdout_read, stdout_read, repl = fake_repl()
+    panel = LineEdit.Prompt("testπ";
+        prompt_prefix="\e[38;5;166m",
+        prompt_suffix=Base.text_colors[:white],
+        on_enter = s->true)
 
-        hp = REPL.REPLHistoryProvider(Dict{Symbol,Any}(:parse => panel))
-        search_prompt, skeymap = LineEdit.setup_prefix_keymap(hp, panel)
-        REPL.history_reset_state(hp)
+    hp = REPL.REPLHistoryProvider(Dict{Symbol,Any}(:parse => panel))
+    search_prompt, skeymap = LineEdit.setup_prefix_keymap(hp, panel)
+    REPL.history_reset_state(hp)
 
-        panel.hist = hp
-        panel.keymap_dict = LineEdit.keymap(Dict{Any,Any}[skeymap,
-            LineEdit.default_keymap, LineEdit.escape_defaults])
+    panel.hist = hp
+    panel.keymap_dict = LineEdit.keymap(Dict{Any,Any}[skeymap,
+        LineEdit.default_keymap, LineEdit.escape_defaults])
 
-        c = Condition()
-        panel.on_done = (s,buf,ok)->begin
-            if !ok
-                LineEdit.transition(s,:abort)
-            end
-            line = strip(String(take!(buf)))
-            LineEdit.reset_state(s)
-            return notify(c,line)
+    c = Condition()
+    panel.on_done = (s,buf,ok)->begin
+        if !ok
+            LineEdit.transition(s,:abort)
         end
-
-        repltask = @async Base.REPL.run_interface(repl.t,
-                              LineEdit.ModalInterface([panel,search_prompt]))
-
-        write(stdin_write,"a\n")
-        @test wait(c) == "a"
-        # Up arrow enter should recall history even at the start
-        write(stdin_write,"\e[A\n")
-        @test wait(c) == "a"
-        # And again
-        write(stdin_write,"\e[A\n")
-        @test wait(c) == "a"
-        # Close REPL ^D
-        write(stdin_write, '\x04')
-        wait(repltask)
+        line = strip(String(take!(buf)))
+        LineEdit.reset_state(s)
+        return notify(c,line)
     end
+
+    repltask = @async Base.REPL.run_interface(repl.t, LineEdit.ModalInterface([panel,search_prompt]))
+
+    write(stdin_write,"a\n")
+    @test wait(c) == "a"
+    # Up arrow enter should recall history even at the start
+    write(stdin_write,"\e[A\n")
+    @test wait(c) == "a"
+    # And again
+    write(stdin_write,"\e[A\n")
+    @test wait(c) == "a"
+    # Close REPL ^D
+    write(stdin_write, '\x04')
+    wait(repltask)
 end
 
 ccall(:jl_exit_on_sigint, Void, (Cint,), 1)

